components/api/src/main/java/com/hotels/styx/api/cookies/RequestCookie.java

```diff
@@ -0,0 +1,149 @@
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.api.cookies;
+
+import com.google.common.base.Objects;
+import io.netty.handler.codec.http.cookie.ClientCookieEncoder;
+import io.netty.handler.codec.http.cookie.Cookie;
+import io.netty.handler.codec.http.cookie.DefaultCookie;
+import io.netty.handler.codec.http.cookie.ServerCookieDecoder;
+
+import java.util.Collection;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.hotels.styx.api.common.Strings.quote;
+import static java.util.Collections.emptySet;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * Represents an HTTP cookie as sent in the {@code Cookie} header of an HTTP request.
+ */
+public final class RequestCookie {
+    private final String name;
+    private final String value;
+    private final int hashCode;
+
+    /**
+     * Constructs a cookie with a name, value and attributes.
+     *
+     * @param name  cookie name
+     * @param value cookie value
+     */
+    private RequestCookie(String name, String value) {
+        checkArgument(!isNullOrEmpty(name), "name cannot be null or empty");
+        requireNonNull(value, "value cannot be null");
+        this.name = name;
+        this.value = value;
+        this.hashCode = Objects.hashCode(name, value);
+    }
+
+    /**
+     * Constructs a cookie with a name, value and attributes.
+     *
+     * @param name  cookie name
+     * @param value cookie value
+     * @return a cookie
+     */
+    public static RequestCookie requestCookie(String name, String value) {
+        return new RequestCookie(name, value);
+    }
+
+    /**
+     * Decodes a "Cookie" header value into a set of {@link RequestCookie} objects.
+     *
+     * @param headerValue "Cookie" header value
+     * @return cookies
+     */
+    public static Set<RequestCookie> decode(String headerValue) {
+        if (headerValue == null) {
+            return emptySet();
+        }
+
+        return ServerCookieDecoder.LAX.decode(headerValue).stream()
+                .map(RequestCookie::convert)
+                .collect(toSet());
+    }
+
+    /**
+     * Encodes a collection of {@link RequestCookie} objects into a "Cookie" header value.
+     *
+     * @param cookies cookies
+     * @return "Cookie" header value
+     */
+    public static String encode(Collection<RequestCookie> cookies) {
```

Comment:
According to the RFC, servers should not discard duplicate cookie values based on the order of appearance: https://tools.ietf.org/html/rfc6265#section-4.2.2  But it seems difficult to understand what´s expected (other than returning a list) to do with these duplicate cookies, have we verified what´s the most common approach to handle this_?

Also, I don´t know this method is semantically correct if we intend to discard all duplicate cookies and keep the first one. A Collection could be traversed in LIFO order... We may consider that it still means that the first cookie returned will be kept and all the following cookies with the same name removed.

Reply:
The treatment of duplicate cookies is part of the netty encoding logic - it is not something introduced by Styx.